Add Route53 + CloudFormation + SQS + StepFunction + S3 report generation

theocerutti opened the pull request:
Add report generation for:

- [X] StepFunctions
- [X] SQS
- [X] S3
- [X] Route 53
- [X] CloudFormation

NOTE:
Tagbot Role Stack required more policy to fetch above resources data.
Then, if you want to test this PR NOW it will not work because current TagBot role are not set.
TagbotRole policies to add during deployment:

```yml
AWSTemplateFormatVersion: 2010-09-09
Description: Create a View Only Access role for tagbot (belong to Trackit/MSolutions.io) and outputs the ARN
Resources:
  TagbotCrossAccountRole:
    Type: 'AWS::IAM::Role'
    Properties:
      RoleName: "TagbotViewAccess"
      AssumeRolePolicyDocument:
        Statement:
          - Action: 'sts:AssumeRole'
            Effect: Allow
            Principal:
              AWS: 'arn:aws:iam::394125495069:root'
      ManagedPolicyArns:
        - 'arn:aws:iam::aws:policy/job-function/ViewOnlyAccess'
        - 'arn:aws:iam::aws:policy/AmazonRDSReadOnlyAccess'
        - 'arn:aws:iam::aws:policy/AmazonESReadOnlyAccess'
      Policies:
        - PolicyName: 'tagbot-view-access'
          PolicyDocument:
            Version: '2012-10-17'
            Statement:
              - Effect: Allow
                Action:
                  - 'sqs:ListQueues'
                  - 'sqs:ListQueueTags'
                  - 'states:ListStateMachines'
                  - 'states:ListTagsForResource'
                  - 's3:ListBucket'
                  - 's3:GetBucketTagging'
                  - 's3:GetBucketLocation'
                  - 'route53:ListHostedZones'
                  - 'route53:ListTagsForResource'
                  - 'cloudformation:ListStacks'
                  - 'cloudformation:DescribeStacks'
                Resource: '*'
Outputs:
  TagbotRoleArn:
    Description: The ARN of the role created for the tagbot web application
    Value: !GetAtt 
      - TagbotCrossAccountRole
      - Arn
```

Commit:
Add Route53 + CloudFormation + SQS + StepFunction + S3 report generation

Commit:
update import coding style

Commit:
Update coding style

simonmeyerrr commented on aws/usageReports/route53/dailyReport.go:
```diff
@@ -0,0 +1,106 @@
+		return err
+	}
+	hostedZonesChans := make([]<-chan HostedZone, 0, len(regions))
+	for _, region := range regions {
```
It seems to me that route53 works globally and not by region, to be confirmed and corrected if this is the case.

GabrielRavier replied:
Can confirm that Route53 has at the very least 2 regions, with the second one being in China, as can be seen on this page: https://docs.aws.amazon.com/general/latest/gr/r53.html

(meaning route53 must logically work by region)

simonmeyerrr commented on aws/usageReports/s3/dailyReport.go:
```diff
@@ -0,0 +1,111 @@
+		return err
+	}
+	bucketChans := make([]<-chan Bucket, 0, len(regions))
+	for _, region := range regions {
```
There is probably a way to optimize as ListBucket list buckets from all regions.

GabrielRavier replied:
https://github.com/trackit/trackit/pull/318/commits/de0d7b0608609802d8ff0e60f1d46bb75a640a86 divides the amount of API calls by (approx) the amount of regions

simonmeyerrr commented on aws/usageReports/stepfunction/utils.go:
```diff
@@ -0,0 +1,132 @@
+	"time"
+
+	"github.com/trackit/jsonlog"
+	taws "github.com/trackit/trackit/aws"
```
missing empty line

GabrielRavier replied:
Should be fixed now

simonmeyerrr commented on db/migration/0053_add_resources_errors_to_account_update_job.sql:
```diff
@@ -0,0 +1,19 @@
+ALTER TABLE aws_account_update_job ADD s3Error VARCHAR(255) NOT NULL DEFAULT "";
+ALTER TABLE aws_account_update_job ADD sqsError VARCHAR(255) NOT NULL DEFAULT "";
+ALTER TABLE aws_account_update_job ADD cloudFormationError VARCHAR(255) NOT NULL DEFAULT "";
+ALTER TABLE aws_account_update_job ADD route53Error VARCHAR(255) NOT NULL DEFAULT "";
```
missing new line

GabrielRavier replied:
Should be fixed now

simonmeyerrr commented on db/schema.sql:
```diff
@@ -1207,3 +1207,23 @@ ALTER TABLE tagbot_user ADD stripe_payment_method_identifier VARCHAR(255) NOT NU
+ALTER TABLE aws_account_update_job ADD s3Error VARCHAR(255) NOT NULL DEFAULT "";
+ALTER TABLE aws_account_update_job ADD sqsError VARCHAR(255) NOT NULL DEFAULT "";
+ALTER TABLE aws_account_update_job ADD cloudFormationError VARCHAR(255) NOT NULL DEFAULT "";
+ALTER TABLE aws_account_update_job ADD route53Error VARCHAR(255) NOT NULL DEFAULT "";
```
missing new line

GabrielRavier replied:
Should be fixed now

simonmeyerrr commented on server/taskProcessAccount.go:
```diff
@@ -23,8 +23,8 @@ import (
 	"time"
 
 	"github.com/trackit/jsonlog"
-
```
put again this empty line please

GabrielRavier replied:
Should be fixed now

simonmeyerrr commented on tagging/stepfunction/process.go:
```diff
@@ -0,0 +1,82 @@
+		ResourceID:   source.Step.Name,
+		ResourceType: resourceTypeString,
+		Region:       source.Step.Region,
+		URL:          fmt.Sprintf(urlFormat, regionForURL, regionForURL, source.Account, source.Step.Name),
```
You have directly ARN in your struct, you don't need to generate it again

GabrielRavier replied:
https://github.com/trackit/trackit/pull/318/commits/45cc4ef6315371132ca8d733e86e89d8414c4831 solves this by using the one from ES

simonmeyerrr reviewed (changes requested):
We should think about a way to update the role easily for users.
Please update policies in `/policies` folder.

GabrielRavier committed:
Merge branch 'stg' into feat/316-resources-report

Update so we can build feat/316-resources-report again

GabrielRavier committed:
Merge branch 'stg' into feat/316-resources-report
Update to latest stg to get ulimits fix

GabrielRavier committed:
Made corrections w.r.t. spacing

GabrielRavier committed:
Updated policies/ files for changes in feat/316-resources-report

GabrielRavier committed:
Noted potential optimization of route53 tag gathering

GabrielRavier committed:
Added note on why we iterate over all regions when doing S3 stuff

GabrielRavier committed:
update-tags: Fixed rare hash collision w.r.t. buckets

Sometimes, buckets will have the same ID, but have different resource
types. Previously, this would result in one of them being randomly discarded.

GabrielRavier committed:
stepfunction tagging: Avoid needlessly manually replacating the ARN

Much simpler to just get it from ES

GabrielRavier committed:
S3: Optimized amount of API calls as much as possible

Originally, this opens connections for every single region, every
single one of which will:
- Do ListBuckets (which will list all buckets from all regions)
- Call GetBucketLocation on every bucket to determine whether it's
  from the right region
- Call GetBucketTagging if the bucket is from the right region (since
  you need to call GetBucketTagging from the corresponding region)

We now instead:
- Do ListBuckets once
- Call GetBucketLocation on every single bucket and determine to which
  region it corresponds, and create a map of regions with an
  associated array of buckets to handle
- Open a connection for every region for which there are buckets,
  which will each do GetBucketTagging for the corresponding buckets so
  it's called from the right region

This approximatively divides the amount of API calls by the amount of
regions

GabrielRavier committed:
S3 usage reports: Avoid terminal flood when getting tags

GabrielRavier committed:
Merge remote-tracking branch 'origin/stg' into feat/316-resources-report

J3suis commented on aws/usageReports/cloudformation/metricsStats.go:
```diff
@@ -0,0 +1,46 @@
+		logger.Error("Failed to get CloudFormation tags", err.Error())
+		return res
+	}
+
```
This empty line must be here ?

GabrielRavier replied:
It's not mandatory or anything, but it makes the code easier to read IMO

EthanHusser commented on aws/usageReports/cloudformation/metricsStats.go:
```diff
@@ -0,0 +1,46 @@
+		logger.Error("Failed to get CloudFormation tags", err.Error())
+		return res
+	}
+
```
Maybe just remove this empty line

GabrielRavier replied:
Already mentioned by Paul

J3suis reviewed (approved):
Everything seems Fine, but there is a lot of line skipping

EthanHusser reviewed (approved):
Nothing to say about the code, just pay attention to some places where there are empty lines.
